src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -187,6 +201,163 @@ class UserService(protected val userInfo: UserInfo, dataSource: DataSource, prot
+        containerDAO.authDAO.loadUserByEmail(user, txn).getOrElse(return false)
+      }
+      memberList.subGroupEmails.foreach { subGroup =>
+        containerDAO.authDAO.loadGroupByEmail(subGroup, txn).getOrElse(return false)
+      }
+      true
+    }
+  }
+
+  def addGroupMembers(groupRef: RawlsGroupRef, memberList: RawlsGroupMemberList) = {
+    asAdmin {
+      dataSource.inFutureTransaction() { txn =>
+        if (!allMembersExist(memberList))
+          Future.successful(RequestComplete(StatusCodes.NotFound, "Not all members are registered. Please ensure that all users/groups exist"))
+        else {
+          val addMap = memberList.userEmails.map { user =>
+            val theUser = containerDAO.authDAO.loadUserByEmail(user, txn).get
+            val addTry = toFutureTry(gcsDAO.addMemberToGoogleGroup(groupRef, Left(theUser)))
+            addTry.map(Left(theUser) -> _)
+          } ++ memberList.subGroupEmails.map { subGroup =>
+            val theGroup = containerDAO.authDAO.loadGroupByEmail(subGroup, txn).get
+            val addTry = toFutureTry(gcsDAO.addMemberToGoogleGroup(groupRef, Right(theGroup)))
+            addTry.map(Right(theGroup) -> _)
+          }
+
+          val list: Future[Seq[Try[Either[RawlsUser, RawlsGroup]]]] = Future.sequence(addMap) map { pairs =>
+            pairs.map { case (member: Either[RawlsUser, RawlsGroup], result: Try[Either[RawlsUser, RawlsGroup]]) =>
+              result match {
+                case Success(_) => Success(member)
+                case Failure(f) => member match {
+                  case Left(theUser) => Failure(new RawlsException(s"Could not add user ${theUser.userEmail.value}", f))
+                  case Right(theGroup) => Failure(new RawlsException(s"Could not add group ${theGroup.groupEmail.value}", f))
+                }
+              }
+            }
+          }
+
+          list.map { tries =>
+            val exceptions = tries.collect {
+              case Failure(t) => t
+            }
+
+            val successfulUsers = tries.collect {
+              case Success(Left(member)) => RawlsUser.toRef(member)
+            }.toSet
+            val successfulGroups = tries.collect {
+              case Success(Right(member)) => RawlsGroup.toRef(member)
+            }.toSet
+
+            val group = containerDAO.authDAO.loadGroup(groupRef, txn).getOrElse(throw new RawlsException("Unable to load group"))
+            containerDAO.authDAO.saveGroup(group.copy(users = (group.users ++ successfulUsers), subGroups = (group.subGroups ++ successfulGroups)), txn)
+            if (exceptions.isEmpty)
+              RequestComplete(StatusCodes.OK)
+            else
+              RequestComplete(ErrorReport(StatusCodes.BadRequest, "Unable to add the following member(s)", exceptions.map(ErrorReport(_))))
+          }
+        }
+      }
+    }
+  }
+
+  def removeGroupMembers(groupRef: RawlsGroupRef, memberList: RawlsGroupMemberList) = {
```

Comment:
is there any commonality between add and remove that can be factored out... the code looks virtually identical